ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -269,3 +269,308 @@ let apply_guest_agent_config ~__context =
+              + (day_configed_int * 24 * 60 * 60)
+              + (hour_configed_int * 60 * 60)
+            in
+            let tm_configed = Unix.gmtime (Int.to_float secs_configed) in
+            tm_configed.Unix.tm_mday
+        in
+        let min_random = secs_random_within_an_hour / 60 in
+        let sec_random = secs_random_within_an_hour mod 60 in
+        let secs_next_run, _ =
+          Unix.mktime
+            {
+              tm_year= year_next_run
+            ; tm_mon= month_next_run
+            ; tm_mday= monthday_next_run
+            ; tm_hour= hour_configed_int
+            ; tm_min= min_random
+            ; tm_sec= sec_random
+            ; tm_wday= 0
+            ; tm_yday= 0
+            ; tm_isdst= false
+            }
+        in
+        debug
+          "[PeriodicUpdateSync] periodic_update_sync_delay monthly, \
+           execute_next_month=%B year_next_run=%d month_next_run=%d \
+           monthday_next_run=%d"
+          execute_next_month (year_next_run + 1900) (month_next_run + 1)
+          monthday_next_run ;
+        secs_next_run -. secs_now
+
+  let update_sync_delay ~__context ~first_run =
+    if Xapi_fist.update_sync_every_ten_minutes () then (
+      debug "[PeriodicUpdateSync] fist update_sync_delay: 10 minutes" ;
+      600.0
+    ) else
+      let delay = periodic_update_sync_delay ~__context ~first_run in
+      debug "[PeriodicUpdateSync] delay for next update sync: %f seconds" delay ;
+      delay
+
+  let rec periodic_update_sync () =
+    Server_helpers.exec_with_new_task "periodic_update_sync" (fun __context ->
+        Xapi_periodic_scheduler.add_to_queue periodic_update_sync_task_name
+          Xapi_periodic_scheduler.OneShot
+          (update_sync_delay ~__context ~first_run:false)
+          periodic_update_sync ;
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            let rec sync_updates_with_retry failed_times =
+              if failed_times < 4 then
+                if failed_times > 0 then
+                  debug
+                    "[PeriodicUpdateSync] retry as pool.sync_updates failed, \
+                     number of retries: %d"
+                    failed_times ;
+              try
+                ignore
+                  (Client.Pool.sync_updates ~rpc ~session_id
+                     ~self:(Helpers.get_pool ~__context)
+                     ~force:false ~token:"" ~token_id:""
+                  )
+              with _ -> sync_updates_with_retry (failed_times + 1)
+            in
+            sync_updates_with_retry 0
+        )
+    )
+
+  let set_enabled ~__context ~value =
+    debug "[PeriodicUpdateSync] set_enabled: %B" value ;
+    if value then (
+      Xapi_periodic_scheduler.remove_from_queue periodic_update_sync_task_name ;
+      Xapi_periodic_scheduler.add_to_queue periodic_update_sync_task_name
+        Xapi_periodic_scheduler.OneShot
+        (update_sync_delay ~__context ~first_run:true)
```

Comment:
Would an immediate run after enabling make more sense? So that there is no need to have a `first_run` in `update_sync_delay`.